doc/ch-run_desc.rst

```diff
@@ -12,62 +12,84 @@ Run command :code:`CMD` in a Charliecloud container using the flattened and
 unpacked image directory located at :code:`NEWROOT`.
 
   :code:`-b`, :code:`--bind=SRC[:DST]`
-    mount :code:`SRC` at guest :code:`DST` (default :code:`/mnt/0`,
-    :code:`/mnt/1`, etc.)
+    Bind-mount :code:`SRC` at guest :code:`DST`. The default destination if
+    not specified is to use the same path as the host; i.e., the default is
+    equivalent to :code:`--bind=SRC:SRC`. Can be repeated.
+
+    If :code:`--write` is given and :code:`DST` does not exist, it will be
+    created as an empty directory. **Be wary** of prior bind mounts, including
+    the defaults, because mount points are created regardless of whether they
```

Comment:
Hmm, this can be a pretty nasty surprise. I wonder if image DST should exist otherwise error.

Reply:
Well the issue is to create `DST` if it doesn't exist. ;)

It's possible to avoid traversing mount points. I don't know how hard that is. That could be used to refuse creating `DST` unless it's actually in the image.

Reply:
OK, I looked into this a bit.

`statvfs(2)` returns a field `f_fsid` that seemed promising, but it's the *filesystem* ID rather than a *mount* ID. E.g., if you bind-mount the same filesystem that holds the image into the image somewhere, you can't use this to find the mount point. So doesn't help unfortunately.

`/proc/self/mountinfo` has the info we need ([man page](https://man7.org/linux/man-pages/man5/proc.5.html)), but it's a text file so requires parsing. Options:

1. Use `libmount`, which is part of `util-linux` and so would be widely available???
    * I was not able to find the `mountinfo` parsing code to see how hairy it was.
    * I'm allergic to dependencies.
2. Parse it ourselves.
    * We'd need to check only one thing, I think: e.g., if `base` is `/foo` and `path` is `/bar/...`, then if `/foo/bar` appears in the file that means *something* in `path` traverses a mount point and we don't care exactly what.
    * `fscanf(3)` looks capable enough for what we need.
    * Paths containing spaces are tricky. Parsing the file itself is unambiguous because spaces are escaped (e.g. `112 24 0:43 / /mnt/ba\040z [...]`), but that means plain string matching won't work. We could just error out if any backslashes are present.
3. Don't add this protection; leave the code as-is and if users create mount points in unexpected places that's their problem.

Opinions?

Reply:
Additional options from chat with @j-ogas:

4. Keep track of bind mounts we do, and use this new list to avoid traversal.

5. Only validate against the default bind mounts.

Reply:
I think I like options 4, 3, and 2 in that order. Let me explain myself and let me know if it seems like I'm missing the point:

Option 4 is nice because it's simple, and it avoids recursive traversals. It's also easy to think about, and requires no external dependencies.

Option 3 is fine for projects I care about, since I think we generally avoid spaces anyway. If a user specifies that they want to bind to an existing directory and over-write the content there, they should be able to do that. I believe they should be trusted to understand that, considering this is how `mount` works by default.

Option 2 would be ok if you felt like more checking/validation is appropriate for paths containing spaces, but paths containing spaces just seem bad and I never use them.

I do not like option 5 because I expect that I will always be using Charliecloud with Spack environments, and if a Spack environment says a path is in `/tmp`, I want to be able to use that without modification. These environments can be difficult to configure, and may represent many hours of work by a team of people, and not necessarily the same person/people who are build the container. Unless binding to a directory explicitly harms Charliecloud, I think I would like the flexibility to do that, and accept the consequences. Informative errors would be important if a bind is rejected.

I don't have strong feelings about option 1, but I generally agree that minimizing dependencies is good.

Reply:
Thanks Quincy. Re. Option 4, under that proposal `ch-run` would refuse to create mount points in `/tmp` unless you also said `--private-tmp`.

Most of these also have error/warn variants, e.g. refuse to run or just warn about it.

Regarding the cost of an error here, it is directories created in unexpected places the user has write access too. That seems relatively low.

Reply:
We selected Option 4.